stream/src/public_key/connection.rs

```diff
@@ -13,13 +13,60 @@ use commonware_cryptography::Scheme;
 use commonware_macros::select;
 use commonware_runtime::{Clock, Sink, Spawner, Stream};
 use commonware_utils::SystemTimeExt as _;
+use hkdf::Hkdf;
 use rand::{CryptoRng, Rng};
+use sha2::{Digest, Sha256};
 use std::time::SystemTime;
 
 // When encrypting data, an encryption tag is appended to the ciphertext.
 // This constant represents the size of the encryption tag in bytes.
 const ENCRYPTION_TAG_LENGTH: usize = 16;
 
+/// Helper function to derive directional keys using HKDF-SHA256.
+///
+/// Returns the dialer-to-listener and listener-to-dialer ciphers (in that order).
+fn derive_directional_ciphers(
+    shared_secret: &[u8],
+    dialer_transcript: &[u8],
+    listener_transcript: &[u8],
+) -> Result<(ChaCha20Poly1305, ChaCha20Poly1305), Error> {
+    // For ChaCha20Poly1305, key length is 32 bytes.
+    const KEY_LEN: usize = 32;
+
+    // HKDF-Extract: Create a pseudorandom key (PRK)
+    let salt: Option<&[u8]> = None; // TODO: derive salt
+    let prk = Hkdf::<Sha256>::new(salt, shared_secret);
+
+    // Base info for KDF-Expand (protocol name + transcript hash)
+    let mut transcript = Vec::with_capacity(dialer_transcript.len() + listener_transcript.len());
+    transcript.extend_from_slice(dialer_transcript);
+    transcript.extend_from_slice(listener_transcript);
+    let transcript_hash = Sha256::digest(&transcript);
```

Comment:
See if `Sha256` itself will allow for adding the transcripts directly rather than first copying into an intermediate `Vec`